dump-c: fix output of _Bool-typed constants

tautschnig opened the pull request:
Only __CPROVER_bool-typed constants can be used with exprt::is_true.
exprt::is_one's documentation promised that ID_c_bool was supported,
which wasn't actually the case. This commit fixes this and makes use of
is_one to fix the bug in dump-c.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
dump-c: use a custom expr2c configuration

This avoids working around expr2c internals.

tautschnig committed:
dump-c: fix output of _Bool-typed constants

Only __CPROVER_bool-typed constants can be used with exprt::is_true.
exprt::is_one's documentation promised that ID_c_bool was supported,
which wasn't actually the case. This commit fixes this and makes use of
is_one to fix the bug in dump-c.

hannes-steffenhagen-diffblue commented on regression/goto-instrument/dump-bool/test.desc:
```diff
@@ -0,0 +1,12 @@
+irep
+--
+Dump-c previously always generated 0 for _Bool or __CPROVER_bool-typed
+constants. This test ensures that the value "1" is correctly generated.
```
I don't understand how this test tests this?

martin-cs replied:
Yeah... I'm not seeing it either.  Not quite sure where the `irep` is coming from.

tautschnig replied:
Test description expanded: the previous (buggy) output made dump-c produce v = 0 instead of v = 1.

hannes-steffenhagen-diffblue commented on src/goto-instrument/dump_c.cpp:
```diff
@@ -1521,22 +1521,48 @@ void dump_ct::cleanup_type(typet &type)
 
+static expr2c_configurationt expr2c_configuration()
+{
+  static expr2c_configurationt configuration{
```
IMHO would be better written as a series of assignments, then you don't need the comments.

hannes-steffenhagen-diffblue commented on src/goto-instrument/dump_c_class.h:
```diff
@@ -156,7 +152,7 @@ class dump_ct
   const namespacet ns;
   const dump_c_configurationt dump_c_config;
-  std::unique_ptr<languaget> language;
+  const irep_idt mode;
```
This change should be a separate commit with an explanation for why it's done. It's not obvious to me that this is an improvement, and in any case it doesn't seem to be directly related to fixing anything with the output of booleans? Not saying I'm against this, just that the commit should do a better job of explaining it.

tautschnig replied:
Broken out into a second commit.

martin-cs commented on src/goto-instrument/dump_c.cpp:
```diff
@@ -526,7 +526,7 @@ void dump_ct::convert_compound(
         struct_body << "long long int " << comp_name
           << " : " << t.get_width();
-      else if(language->id()=="cpp")
+      else if(mode == ID_cpp)
```
Doesn't seem to correspond with the PR message.  I'm fine with them going in as a minor fix but just wanted to flag it in case it was a mistake.

tautschnig committed:
dump-c: handle further byte_update union initializers

The CSmith test generated with random seed 1614606277 demonstrated that
dump-c's handling still didn't cover all cases of byte_update expressions
introduced by the C front-end, specifically missing the case where a struct was
part of the union.
